### PR TITLE
Fix mrc bugs

### LINKED ISSUE #71: Error in reading 4DSTEM .mrc files from Velox
Dear HyperSpy community,
I would like to report an error in loading .mrc files created using Velox. It was discussed previously in issue hyperspy/rosettasciio#130. The error is the same as in the issue hyperspy/rosettasciio#130, but I listed it below for ease of reference.  In my case the situation is less easy to solve, because the .mrc files contain 4DSTEM data, which you cannot easily save as different format in Velox. The workaround I use is to load the dataset to ImageJ and save it as a Tiff stack which you can load into hyperspy. This approach requires you to have enough memory to load the whole dataset, but these files can easily exceed 100 GB, which can be problematic if you do not have workstation with enough memory. Therefore, it would be very useful to be able to read the file using hs.load and take advantage of lazy loading. 
There is a full description of their .mrc header in Velox manual, which is unfortunately confidential, but I can ask them for permission to share it.

Petr

#### Error
```
WARNING:hyperspy.io_plugins.mrc:There was a problem reading the extended header
ERROR:hyperspy.io:If this file format is supported, please report this error to the HyperSpy developers.
Output exceeds the [size limit](command:workbench.action.openSettings?[). Open the full output data [in a text editor](command:workbench.action.openLargeOutput?d72db32d-806f-4926-ae72-9110b3125e9f)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
c:\Users\Petr\Desktop\recycled battery\edx.ipynb Cell 2 in <cell line: 1>()
----> [1](vscode-notebook-cell:/c%3A/Users/Petr/Desktop/recycled%20battery/edx.ipynb#X66sZmlsZQ%3D%3D?line=0) a = hs.load('D:/Práce/Experimenty/raw data/20221014 Camera 91 mm Ceta 1520-e4df9eac320047e6b030968f25823eb0.mrc')

File c:\Users\Petr\anaconda3\envs\hyperspy171pyxem0142\lib\site-packages\hyperspy\io.py:454, in load(filenames, signal_type, stack, stack_axis, new_axis_name, lazy, convert_units, escape_square_brackets, stack_metadata, load_original_metadata, show_progressbar, **kwds)
    451         objects.append(signal)
    452 else:
    453     # No stack, so simply we load all signals in all files separately
--> 454     objects = [load_single_file(filename, lazy=lazy, **kwds)
    455                for filename in filenames]
    457 if len(objects) == 1:
    458     objects = objects[0]

File c:\Users\Petr\anaconda3\envs\hyperspy171pyxem0142\lib\site-packages\hyperspy\io.py:454, in <listcomp>(.0)
    451         objects.append(signal)
    452 else:
    453     # No stack, so simply we load all signals in all files separately
--> 454     objects = [load_single_file(filename, lazy=lazy, **kwds)
    455                for filename in filenames]
    457 if len(objects) == 1:
    458     objects = objects[0]

File c:\Users\Petr\anaconda3\envs\hyperspy171pyxem0142\lib\site-packages\hyperspy\io.py:513, in load_single_file(filename, **kwds)
    506     raise ValueError(
...
--> 155     f.seek(1024 + std_header['NEXT'])
    156     fei_header = None
    157 NX, NY, NZ = std_header['NX'], std_header['NY'], std_header['NZ']

TypeError: only integer scalar arrays can be converted to a scalar index
```

### PR DESCRIPTION
**Pull request by ericpre**
Fix #71, #91, #93, #96, #130 and some of #60.

Currently including #129 to test updating registry from pull request

### Progress of the PR
- [x] Fix reading standard header for non-FEI mrc (#96),
- [x] Fix reading unsigned 16 bits data (#71)
- [x] Fix reading scale for non-FEI mrc (#91)
- [x] Workaround for setting scales when z scale is 0 of mrc file generate by Velox
- [n/a] update docstring (if appropriate),
- [n/a] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/rosettasciio/blob/main/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests,
- [x] ready for review.

### CODE REVIEW

**Comment by ericpre** [comment]
pre-commit.ci autofix

File: rsciio/mrc/_api.py [hunk 1]
```diff
@@ -188,20 +193,20 @@ def file_reader(filename, lazy=False, mmap_mode=None, endianess="<", **kwds):
     if fei_header is None:
         # The scale is in Amstrongs, we convert it to nm
```
Comment (jlaehne) [inline rsciio/mrc/_api.py:194]:
```suggestion
        # The scale is in Angstroms, we convert it to nm
```

:-)

**Comment by jlaehne** [comment]
DOes this PR also fix  #130?

**Comment by jlaehne** [comment]
@ericpre can you rebase to throw out the unrelated changes and ease the review.

**Comment by ericpre** [comment]
There is one more that I need to do here is add an option to specify the navigation shape, because in case of 4D-STEM data saved by Velox, the data seems to be saved as a stack even if the MRC2014 specification supports it - this is consistent with the way [mrcfile](https://mrcfile.readthedocs.io) reads the data.

**Comment by ericpre** [comment]
> DOes this PR also fix #130?

Yes, it does! :)

**Comment by ericpre** [comment]
The failure of the "Test Packaging" build is expected because it downloads the test files from the `main` branch and the file are not there, not until the PR is merged.

File: docs/supported_formats/mrc.rst [hunk 2]
```diff
@@ -10,6 +10,12 @@ partly support FEI's custom header. We do not provide writing features for this
 format, but, as it is an open format, we may implement this feature in the
 future on demand.
 
+.. Note::
+    When reading 4D-STEM data saved by the Velox software, the data are read as a stack
+    of diffraction pattern, but the ``navigation_shape`` argument can be used to
```
Comment (jlaehne) [inline docs/supported_formats/mrc.rst:15]:
```suggestion
    of diffraction patterns, but the ``navigation_shape`` argument can be used to
```

**Comment by jlaehne** [comment]
Would it make sense to add the non-FEI test file from #96 to the tests?
(though it seems that the so far not tested parts of the code seem to be related to FEI headers and this extra file might not give any added value)

**Comment by ericpre** [comment]
> Would it make sense to add the non-FEI test file from #96 to the tests? (though it seems that the so far not tested parts of the code seem to be related to FEI headers and this extra file might not give any added value)

Yes, `mrc` files generated by Velox seems to use the "non-FEI" files code path, which is covered by the other files. The file provided in #96 are larger and doesn't seem to increase coverage.

**Comment by ericpre** [comment]
While I was at it, improve the documentation to close #93! :)

**Comment by jlaehne** [comment]
> > Would it make sense to add the non-FEI test file from #96 to the tests? (though it seems that the so far not tested parts of the code seem to be related to FEI headers and this extra file might not give any added value)
> 
> Yes, `mrc` files generated by Velox seems to use the "non-FEI" files code path, which is covered by the other files. The file provided in #96 are larger and doesn't seem to increase coverage.

Well one of the current test files is 32 MB, so larger than the 4.5 MB one from #96, but I don't know if could replace that one.

File: docs/supported_formats/mrc.rst [hunk 3]
```diff
@@ -3,22 +3,56 @@
...
+
+    import numpy as np
+    from rsciio import mrcz
+
+    data = np.random.randint(100, size=(10, 100, 100)).astype('int16')
+    s = hs.signals.Signal2D(data)
+    s_dict = s.as_dictionary()
+
+    # If the compressor argument, it will not be a mrcz file and 
+    # not a MRC compliant file
```
Comment (jlaehne) [inline docs/supported_formats/mrc.rst:40]:
```suggestion
```

File: docs/supported_formats/mrc.rst [hunk 4]
```diff
@@ -3,22 +3,56 @@
...
+    mrcz.file_writer('test.mrc', s_dict)
+
+Or with using the :py:meth:`hyperspy.signal.BaseSignal.save`, which will pick the
+``mrcz`` plugin automatically:
```
Comment (jlaehne) [inline docs/supported_formats/mrc.rst:44]:
```suggestion
Alternatively, use :py:meth:`hyperspy.signal.BaseSignal.save`, which will pick the
``mrcz`` plugin automatically:
```

File: docs/supported_formats/mrc.rst [hunk 5]
```diff
@@ -3,22 +3,56 @@
...
-The ``.mrc`` format is widely used for tomographic data. Our implementation is based
-on `this specification
-<https://www2.mrc-lmb.cam.ac.uk/research/locally-developed-software/image-processing-software/>`_. We also
-partly support FEI's custom header. We do not provide writing features for this
-format, but, as it is an open format, we may implement this feature in the
-future on demand.
-
-.. Note ::
-    For ``.mrc`` files, the ``file_reader`` takes the ``mmap_mode`` keyword argument
+The ``mrc`` format is widely used for tomographic data. The implementation of this plugin
```
Comment (jlaehne) [inline docs/supported_formats/mrc.rst:6]: We currently write the format extensions with a dot for most file formats, e.g. ``.bcf`` ...

File: docs/supported_formats/mrcz.rst [hunk 6]
```diff
@@ -3,24 +3,31 @@
...
+   * - PyPI
+     - https://pypi.org/project/mrcz
+   * - Citation
+     - https://doi.org/10.1016/j.jsb.2017.11.012
+   * - Preprint
+     - https://www.biorxiv.org/content/10.1101/116533v1
 
 Support for this format is not enabled by default. In order to enable it
-install the `mrcz` and optionally the `blosc` Python packages.
+the `mrcz` library needs to be installed and optionally the `blosc` to use
```
Comment (jlaehne) [inline docs/supported_formats/mrcz.rst:29]:
```suggestion
the `mrcz` library needs to be installed and optionally `blosc` to use
```

File: docs/supported_formats/mrcz.rst [hunk 7]
```diff
@@ -3,24 +3,31 @@
...
+   * - Repository
+     - https://github.com/em-MRCZ
+   * - PyPI
+     - https://pypi.org/project/mrcz
+   * - Citation
+     - https://doi.org/10.1016/j.jsb.2017.11.012
+   * - Preprint
+     - https://www.biorxiv.org/content/10.1101/116533v1
 
 Support for this format is not enabled by default. In order to enable it
```
Comment (jlaehne) [inline docs/supported_formats/mrcz.rst:28]:
```suggestion
Support for this format is not enabled by default. In order to enable it,
```

**Comment by ericpre** [comment]
> Well one of the current test files is 32 MB, so larger than the 4.5 MB one from #96, but I don't know if could replace that one.

Sorry, there is a typo in my message and anyway I wasn't very clear! What I should have say to explain my reasoning is:
- the mrc 4D-STEM test file generated by Velox seems to be as[ small as they can be](https://github.com/hyperspy/rosettasciio/issues/71#issuecomment-1339780899).
- the mrc files generated by IMOD could be smaller and it would be good to ask/get a smaller one and add it later.

**Comment by ericpre** [comment]
I would suggest to add other type of mrc files (one from the type in #91) in another PR, once we have got smallest possible file.

**Review by jlaehne — APPROVED** [review]

**Pull request merged**